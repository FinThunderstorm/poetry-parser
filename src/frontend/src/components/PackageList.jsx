import React from 'react'
import { Link, Outlet, useParams } from 'react-router-dom'
import { useSelector } from 'react-redux'
import { Alert, Button, Grid, Paper, Stack, Typography } from '@mui/material'

/** PackageList gives list of all packages and offers links to more information
 */
const PackageList = () => {
    const packages = useSelector((state) => state.packages.packages)
    const status = useSelector((state) => state.packages.status)
    const lockVersion = useSelector((state) => state.packages.lockVersion)
    const params = useParams()

    if (status === 'idle') {
        return <></>
    }

    return (
        <>
            <Grid item xs={3}>
                <Paper
                    elevation={6}
                    sx={{ height: '80vh', p: 2, overflowY: 'scroll' }}
                >
                    <Typography variant="button" sx={{ fontSize: '1.5rem' }}>
                        Packages:
                    </Typography>
                    <Stack spacing={1} alignItems="stretch">
                        {status === 'loading' && (
                            <Alert severity="info">Loading...</Alert>
                        )}
                        {status === 'failed' && (
                            <Alert severity="error">
                                Error while parsing packages.
                            </Alert>
                        )}
                        {status === 'succeeded' &&
                            lockVersion !== '1.1' &&
                            lockVersion !== undefined && (
                                <Alert severity="warning">
                                    Currently supported poetry.lock -file
                                    version is 1.1, other versions might have
                                    errors in parsing.
                                </Alert>
                            )}
                        {status === 'succeeded' &&
                            Object.values(packages)
                                .sort((a, b) => {
                                    const nameA = a.name.toLowerCase()
                                    const nameB = b.name.toLowerCase()
                                    if (nameA < nameB) {
                                        return -1
                                    }
                                    if (nameA > nameB) {
                                        return 1
                                    }
                                    return 0
                                })
                                .map((pack) => (
                                    <Button
                                        variant={
                                            params.name === pack.name
                                                ? 'contained'
                                                : 'outlined'
                                        }
                                        component={Link}
                                        key={pack.name}
                                        to={`/packages/${pack.name}`}
                                    >
                                        {pack.name}
                                    </Button>
                                ))}
                    </Stack>
                </Paper>
            </Grid>
            <Grid item xs={9}>
                <Outlet />
            </Grid>
        </>
    )
}

export default PackageList
