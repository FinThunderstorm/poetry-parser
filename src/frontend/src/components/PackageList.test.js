/* eslint-disable react/prop-types */
import React from 'react'
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom/extend-expect'
import { render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { configureStore } from '@reduxjs/toolkit'
import PackageList from './PackageList'

import packagesReducer from '../reducers/packagesSlice'

describe('PackageList', () => {
    const preloadedState = {
        packages: {
            packages: {
                atomicwrites: {
                    name: 'atomicwrites',
                    description: 'Atomic file writes.',
                    optional: 'false',
                    pythonVersions: [
                        '>=2.7',
                        '!=3.0.*',
                        '!=3.1.*',
                        '!=3.2.*',
                        '!=3.3.*',
                    ],
                    dependencies: {},
                    reverseDependencies: [
                        {
                            name: 'pytest',
                            installed: true,
                        },
                    ],
                },
                cachecontrol: {
                    name: 'cachecontrol',
                    description: 'httplib2 caching for requests',
                    optional: 'false',
                    pythonVersions: ['>=3.6'],
                    dependencies: {
                        lockfile: {
                            name: 'lockfile',
                            category: 'filecache',
                            optional: true,
                            source: 'extras',
                            installed: true,
                        },
                        msgpack: {
                            name: 'msgpack',
                            optional: false,
                            source: 'dependencies',
                            installed: true,
                        },
                        requests: {
                            name: 'requests',
                            optional: false,
                            source: 'dependencies',
                            installed: true,
                        },
                        redis: {
                            name: 'redis',
                            category: 'redis',
                            optional: true,
                            source: 'extras',
                            installed: false,
                        },
                    },
                    reverseDependencies: [],
                },
            },
            status: 'succeeded',
            error: null,
        },
    }

    const PackageListWrapper = ({ url }) => {
        const store = configureStore({
            reducer: {
                packages: packagesReducer,
            },
            preloadedState,
        })
        return (
            <MemoryRouter initialEntries={[url]}>
                <Provider store={store}>
                    <Routes>
                        <Route path="/">
                            <Route path="packages" element={<PackageList />} />
                        </Route>
                    </Routes>
                </Provider>
            </MemoryRouter>
        )
    }

    test('PackageList works with reverse dependencies (atomicwrites)', () => {
        render(<PackageListWrapper url="/packages" />)

        const name = screen.getByText('Packages:', {
            exact: false,
        })
        const atomicwrites = screen.getByText('atomicwrites', {
            exact: false,
        })
        const cachecontrol = screen.getByText('cachecontrol', {
            exact: false,
        })
        expect(name).toBeDefined()
        expect(atomicwrites).toBeDefined()
        expect(cachecontrol).toBeDefined()

        const tree = renderer
            .create(<PackageListWrapper url="/packages" />)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
