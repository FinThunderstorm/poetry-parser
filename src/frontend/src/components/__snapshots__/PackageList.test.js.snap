// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageList PackageList works with reverse dependencies (atomicwrites) 1`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-1o7apob-MuiGrid-root"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 css-cxfaie-MuiPaper-root"
    >
      <span
        className="MuiTypography-root MuiTypography-button css-1sqoee-MuiTypography-root"
      >
        Packages:
      </span>
      <div
        className="css-1okj3ks-MuiStack-root"
      >
        <a
          className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          href="/packages/atomicwrites"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          atomicwrites
        </a>
        <a
          className="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          href="/packages/cachecontrol"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
        >
          cachecontrol
        </a>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10 css-17p3wh3-MuiGrid-root"
  />
</div>
`;
